src/Diagnostics.DataProviders/DataProviders/GeoMasterDataProvider.cs

```diff
@@ -608,22 +611,35 @@ public Task<T> InvokeDaasExtension<T>(string subscriptionId, string resourceGrou
 
         private Task<R> HttpGet<R>(string path, string queryString = "", string apiVersion = GeoMasterConstants.August2016Version, CancellationToken cancellationToken = default(CancellationToken))
         {
-            return PerformHttpRequest<R>(HttpMethod.Get, path, queryString, null, apiVersion, cancellationToken);
+            //return PerformHttpRequest<R>(HttpMethod.Get, path, queryString, null, apiVersion, cancellationToken);
+            return PerformHttpRequestWithRetry<R>(HttpMethod.Get, path, queryString, null, apiVersion, cancellationToken);
         }
 
         private Task<R> HttpPost<R, T>(string path, T content = default(T), string queryString = "", string apiVersion = GeoMasterConstants.August2016Version, CancellationToken cancellationToken = default(CancellationToken))
         {
             var body = JsonConvert.SerializeObject(content);
-            return PerformHttpRequest<R>(HttpMethod.Post, path, queryString, body, apiVersion, cancellationToken);
+            //return PerformHttpRequest<R>(HttpMethod.Post, path, queryString, body, apiVersion, cancellationToken);
+            return PerformHttpRequestWithRetry<R>(HttpMethod.Post, path, queryString, body, apiVersion, cancellationToken);
         }
 
+        private Task<R> PerformHttpRequestWithRetry<R>(HttpMethod method, string path, string queryString, string content, string apiVersion, CancellationToken cancellationToken)
+        {
+            //Save params  into tuple temporarly to so to call PerformHttpRequest method without passing params for using RetryHelper
+            HttpRequestTuple = new Tuple<HttpMethod, string, string, string, string, CancellationToken>(method, path, queryString, content, apiVersion, cancellationToken);
```

Comment:
is this multiple execution safe approach?
what would happen if in the same detector, I do the following at the same time:

<pre>
var task1 = dp.Geomaster.Execute..Reuest(method1, path1, .....);
var task2 = dp.Geomaster.Execute..Reuest(method2, path2, .....);

await Task.WhenAll(task1, task2);
</pre>

Since for one detector API call, we only create one object of each data provider (as far as I remember), isnt there a possibility that this HttpRequestTuple is not thread safe and will get overwritten

Reply:
Good point, do you think it will be more thread safe if I change the current logic to store task id with tuple into a concurrrent dictionary. Once task is done, remove entry from dictionary ?

Reply:
Hey @XiaoxuHU , if I am understanding this correctly. The reason we're going with the Tuple approach is because `RetryHelper` currently does not accept parameters for the `Func` that we pass?

Reply:
Yes, @hforeste. You are right.
Basically, I tried to use tuple is to make `Func` in `RetryHelper`  more generic since different data providers have different parmas for their method which pass into `Func`. 

But seems like it might be got overwritten since Tuple is not thread safe. I am thinking of using ConcurrentHashMap to store an id and tuple params for each func instead of tuple to make it safe. But one thing it is that which stuff can be the id. I have tried the task id where task is the return value after execute Func in RetryHelper. But this approach is not working since I am not able to get task id before func get invoked(I need to get id before func get invoke in RetryHelper so to put id and tuple into map) . Also I tried Func hashcode, but if call same function twice with different params, their hash code will be the same and map value will get overwritten.

Any ideas on how I can fix this issue, or do you think any other approaches I can implement ?

Thanks

Reply:
@ShekharGupta1988 and @hforeste , I updated code to use Concurrent Dictionary to store arguments from `Func` which goes to RetryHelper for thread safe

Reply:
> Any ideas on how I can fix this issue, or do you think any other approaches I can implement ?

Hey @XiaoxuHU , one approach is creating overloaded methods of that RetryHelper that take Func with parameters. .NET does this with Func https://docs.microsoft.com/en-us/dotnet/api/system.func-2?view=net-5.0 
![image](https://user-images.githubusercontent.com/917998/110988295-3586fe00-8325-11eb-8c3c-c40e01eedbb5.png)

Reply:
> > Any ideas on how I can fix this issue, or do you think any other approaches I can implement ?
> 
> Hey @XiaoxuHU , one approach is creating overloaded methods of that RetryHelper that take Func with parameters. .NET does this with Func https://docs.microsoft.com/en-us/dotnet/api/system.func-2?view=net-5.0
> ![image](https://user-images.githubusercontent.com/917998/110988295-3586fe00-8325-11eb-8c3c-c40e01eedbb5.png)

@hforeste, the thing is for `Func` in RetryHelper is we only know its return type is `TResult` but we don't know exact types for **input prams** since different DPs will pass different signature of their functions got invoked in RetryHelper